File: airbyte-data/src/main/java/io/airbyte/data/services/OAuthService.java
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.data.services;

import io.airbyte.config.DestinationOAuthParameter;
import io.airbyte.config.SourceOAuthParameter;
import io.airbyte.data.exceptions.ConfigNotFoundException;
import java.io.IOException;
import java.util.Optional;
import java.util.UUID;

/**
 * OAuth Service.
 */
public interface OAuthService {

  void writeSourceOAuthParam(SourceOAuthParameter sourceOAuthParameter) throws IOException;

  Optional<SourceOAuthParameter> getSourceOAuthParameterWithSecretsOptional(UUID workspaceId, UUID sourceDefinitionId)
      throws IOException, ConfigNotFoundException;

  Optional<SourceOAuthParameter> getSourceOAuthParameterOptional(UUID workspaceId, UUID sourceDefinitionId)
      throws IOException;

  Optional<SourceOAuthParameter> getSourceOAuthParameterOptional(UUID workspaceId, UUID organizationId, UUID sourceDefinitionId)
      throws IOException;

  Optional<SourceOAuthParameter> getSourceOAuthParamByDefinitionIdOptional(Optional<UUID> workspaceId,
                                                                           Optional<UUID> organizationId,
                                                                           UUID sourceDefinitionId)
      throws IOException;

  void writeDestinationOAuthParam(DestinationOAuthParameter destinationOAuthParameter) throws IOException;

  Optional<DestinationOAuthParameter> getDestinationOAuthParameterWithSecretsOptional(UUID workspaceId, UUID destinationDefinitionId)
      throws IOException, ConfigNotFoundException;

  Optional<DestinationOAuthParameter> getDestinationOAuthParameterOptional(UUID workspaceId, UUID destinationDefinitionId)
      throws IOException;

  Optional<DestinationOAuthParameter> getDestinationOAuthParameterOptional(UUID workspaceId, UUID organizationId, UUID destinationDefinitionId)
      throws IOException;

  Optional<DestinationOAuthParameter> getDestinationOAuthParamByDefinitionIdOptional(Optional<UUID> workspaceId,
                                                                                     Optional<UUID> organizationId,
                                                                                     UUID destinationDefinitionId)
      throws IOException;

}

File: airbyte-data/src/main/kotlin/io/airbyte/data/services/OAuthService.kt
/*
 * Copyright (c) 2020-2026 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.data.services

import io.airbyte.config.DestinationOAuthParameter
import io.airbyte.config.SourceOAuthParameter
import java.util.Optional
import java.util.UUID

/**
 * OAuth Service.
 */
interface OAuthService {
  fun writeSourceOAuthParam(sourceOAuthParameter: SourceOAuthParameter)

  fun getSourceOAuthParameterWithSecretsOptional(
    workspaceId: UUID,
    sourceDefinitionId: UUID,
  ): Optional<SourceOAuthParameter>

  fun getSourceOAuthParameterOptional(
    workspaceId: UUID,
    sourceDefinitionId: UUID,
  ): Optional<SourceOAuthParameter>

  fun getSourceOAuthParameterOptional(
    workspaceId: UUID,
    organizationId: UUID,
    sourceDefinitionId: UUID,
  ): Optional<SourceOAuthParameter>

  fun getSourceOAuthParamByDefinitionIdOptional(
    workspaceId: Optional<UUID>,
    organizationId: Optional<UUID>,
    sourceDefinitionId: UUID,
  ): Optional<SourceOAuthParameter>

  /**
   * Deletes the organization-level source OAuth parameter override for the given definition.
   *
   * @return number of rows deleted.
   */
  fun deleteSourceOAuthParamByDefinitionId(
    organizationId: UUID,
    sourceDefinitionId: UUID,
  ): Int

  fun writeDestinationOAuthParam(destinationOAuthParameter: DestinationOAuthParameter)

  fun getDestinationOAuthParameterWithSecretsOptional(
    workspaceId: UUID,
    destinationDefinitionId: UUID,
  ): Optional<DestinationOAuthParameter>

  fun getDestinationOAuthParameterOptional(
    workspaceId: UUID,
    destinationDefinitionId: UUID,
  ): Optional<DestinationOAuthParameter>

  fun getDestinationOAuthParameterOptional(
    workspaceId: UUID,
    organizationId: UUID,
    destinationDefinitionId: UUID,
  ): Optional<DestinationOAuthParameter>

  fun getDestinationOAuthParamByDefinitionIdOptional(
    workspaceId: Optional<UUID>,
    organizationId: Optional<UUID>,
    destinationDefinitionId: UUID,
  ): Optional<DestinationOAuthParameter>

  /**
   * Deletes the organization-level destination OAuth parameter override for the given definition.
   *
   * @return number of rows deleted.
   */
  fun deleteDestinationOAuthParamByDefinitionId(
    organizationId: UUID,
    destinationDefinitionId: UUID,
  ): Int
}
